File: shared/src@jvm/io/github/yuanbaobaoo/petallink/ui/components/mate/MateButton.kt
@file:Suppress("FunctionName")

package io.github.yuanbaobaoo.petallink.ui.components.mate

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.wrapContentSize
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.github.yuanbaobaoo.petallink.ui.components.MateIcon
import io.github.yuanbaobaoo.petallink.ui.theme.BrandActive
import io.github.yuanbaobaoo.petallink.ui.theme.BrandColor
import io.github.yuanbaobaoo.petallink.ui.theme.BrandGradient
import io.github.yuanbaobaoo.petallink.ui.theme.Brand100
import io.github.yuanbaobaoo.petallink.ui.theme.BrandLighter
import io.github.yuanbaobaoo.petallink.ui.theme.DesignTokens
import io.github.yuanbaobaoo.petallink.ui.theme.ErrorColor
import io.github.yuanbaobaoo.petallink.ui.theme.LocalSemanticColors
import io.github.yuanbaobaoo.petallink.ui.theme.LocalReducedMotion

/**
 * 按钮形态（v2：主按钮 / 文字 / 图标 / 图标文字 / 软色）。
 */
enum class MateButtonVariant { PRIMARY, TEXT, ICON, ICON_TEXT, SOFT }

/**
 * 统一按钮（v2 重设计）。
 *
 * 五变体视觉严格遵守 v2 设计系统：
 * - primary：品牌渐变底（135° #4A8BF0→#0053DB），#fff，radius-md(8)，h36，带品牌色柔影；hover 降透明；danger→error 纯色
 * - soft：brand-lighter 浅蓝底，brand 字，radius-md(8)，h36；hover→brand-100
 * - text：透明底，brand 字，radius-sm(5)；hover→brand-lighter；danger color error
 * - icon：透明，32×32 正圆形；hover→bg-fill + text-primary；danger color error
 * - icon-text：透明，h36，radius-md(8)；hover→bg-fill
 *
 * hover 仅背景色过渡，无 ripple（clickable indication=null）；loading 时 primary 显 spinner，其余图标 spin + 禁用点击。
 *
 * @param label 按钮文字（icon 变体可空）
 * @param variant 按钮形态
 * @param onClick 点击回调
 * @param icon 图标 name（可选；primary/text 用 14px，icon-text/soft 16px，icon 18px）
 * @param danger 危险态（红）
 * @param disabled 禁用
 * @param loading 加载中（primary 显 spinner，其余图标 spin + 禁用点击）
 * @param fullWidth 100% 宽（仅 primary/text/soft 生效）
 * @param badge 角标数字（>0 显示，仅 icon/icon-text）
 * @param height 自定义高度（0 用变体默认）
 */
@Composable
fun MateButton(
    label: String? = null,
    variant: MateButtonVariant = MateButtonVariant.PRIMARY,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    icon: String? = null,
    danger: Boolean = false,
    disabled: Boolean = false,
    loading: Boolean = false,
    fullWidth: Boolean = false,
    badge: Int = 0,
    height: Dp = 0.dp,
) {
    val interaction = remember { MutableInteractionSource() }
    val hovered by interaction.collectIsHoveredAsState()
    val pressed by interaction.collectIsPressedAsState()
    val effectiveDisabled = disabled || loading
    val semantic = LocalSemanticColors.current
    val reducedMotion = LocalReducedMotion.current

    // 形状：icon 为正圆，text 为 radius-sm，其余 radius-md
    val shape = when (variant) {
        MateButtonVariant.ICON -> CircleShape
        MateButtonVariant.TEXT -> RoundedCornerShape(5.dp)
        else -> RoundedCornerShape(8.dp)
    }

    // 背景与文字色按变体 × 状态组合计算（无 ripple，纯背景过渡）。
    // primary 返回的 bgColor 仅作 danger/disabled 兜底；正常态用渐变刷。
    val pair: Pair<Color, Color> = when (variant) {
        MateButtonVariant.PRIMARY -> when {
            effectiveDisabled -> semantic.border to Color.White
            danger -> (if (pressed) ErrorColor.copy(alpha = 0.85f) else ErrorColor) to Color.White
            else -> BrandColor to Color.White
        }
        MateButtonVariant.SOFT -> when {
            effectiveDisabled -> semantic.bgFill to semantic.textPlaceholder
            hovered -> Brand100 to BrandColor
            else -> BrandLighter to BrandColor
        }
        MateButtonVariant.TEXT -> when {
            effectiveDisabled -> Color.Transparent to semantic.textPlaceholder
            danger -> (if (hovered) BrandLighter else Color.Transparent) to ErrorColor
            hovered -> BrandLighter to BrandColor
            else -> Color.Transparent to if (danger) ErrorColor else BrandColor
        }
        MateButtonVariant.ICON -> when {
            effectiveDisabled -> Color.Transparent to semantic.textPlaceholder
            hovered -> semantic.bgFill to semantic.textPrimary
            else -> Color.Transparent to if (danger) ErrorColor else semantic.textSecondary
        }
        MateButtonVariant.ICON_TEXT -> when {
            effectiveDisabled -> Color.Transparent to semantic.textPlaceholder
            hovered -> semantic.bgFill to semantic.textPrimary
            else -> Color.Transparent to semantic.textPrimary
        }
    }
    val bgColor = pair.first
    val contentColor = pair.second

    val resolvedHeight = when {
        height > 0.dp -> height
        variant == MateButtonVariant.ICON -> 32.dp
        else -> 36.dp
    }
    val iconSize = when (variant) {
        MateButtonVariant.ICON -> 18.dp
        MateButtonVariant.ICON_TEXT, MateButtonVariant.SOFT -> 16.dp
        else -> if (icon != null) 14.dp else 0.dp
    }
    val horizontalPadding = when (variant) {
        MateButtonVariant.ICON -> 0.dp
        MateButtonVariant.ICON_TEXT -> 14.dp
        MateButtonVariant.TEXT -> 8.dp
        MateButtonVariant.SOFT -> 16.dp
        MateButtonVariant.PRIMARY -> 18.dp
    }

    // primary 正常态（非 danger 非禁用）用品牌渐变 + 品牌色柔影，其余用纯色背景。
    val useGradient = variant == MateButtonVariant.PRIMARY && !danger && !effectiveDisabled
    val base = modifier
        .then(
            if (fullWidth && variant != MateButtonVariant.ICON) Modifier.fillMaxWidth()
            else Modifier.wrapContentSize(),
        )
        .height(resolvedHeight)
        .then(
            if (useGradient) Modifier.shadow(
                elevation = 6.dp,
                shape = shape,
                clip = false,
                ambientColor = BrandColor.copy(alpha = 0.35f),
                spotColor = BrandColor.copy(alpha = 0.35f),
            ) else Modifier,
        )
        .clip(shape)
        .then(
            if (useGradient) Modifier.background(BrandGradient)
            else if (variant == MateButtonVariant.PRIMARY && pressed && !danger && !effectiveDisabled) Modifier.background(BrandActive)
            else Modifier.background(bgColor),
        )
        .alpha(if (effectiveDisabled) 0.5f else 1f)
        .hoverable(interaction)
        .then(
            if (effectiveDisabled) Modifier
            else Modifier.clickable(
                interactionSource = interaction,
                indication = null,
                onClick = onClick,
            ),
        )

    // icon 变体固定正方形（32×32 命中区）。
    val sized = if (variant == MateButtonVariant.ICON) base.size(32.dp) else base
    Row(
        modifier = sized.padding(horizontal = horizontalPadding),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        if (loading && variant == MateButtonVariant.PRIMARY) {
            MateButtonSpinner(size = 16.dp, color = contentColor, reducedMotion = reducedMotion)
            if (label != null) Spacer(Modifier.width(8.dp))
        } else if (icon != null) {
            val spin = loading && variant != MateButtonVariant.PRIMARY && !reducedMotion
            MateIcon(name = icon, size = iconSize, tint = contentColor, spin = spin)
            if (label != null) Spacer(Modifier.width(6.dp))
        }
        if (label != null) {
            val fontSize = if (variant == MateButtonVariant.TEXT) {
                DesignTokens.FONT_BODY_SM.sp
            } else {
                DesignTokens.FONT_BODY.sp
            }
            Text(text = label, color = contentColor, fontSize = fontSize, fontWeight = FontWeight.Medium)
        }
        if (badge > 0 && (variant == MateButtonVariant.ICON || variant == MateButtonVariant.ICON_TEXT)) {
            Box(modifier = Modifier.padding(start = 2.dp)) { MateBadge(count = badge) }
        }
    }
}

/**
 * primary 加载态的小 spinner（16×16，0.8s 线性旋转）。
 */
@Composable
private fun MateButtonSpinner(size: Dp, color: Color, reducedMotion: Boolean) {
    if (reducedMotion) {
        // 降级为静态半透明圆，提示「加载中」但不旋转。
        Box(Modifier.size(size).clip(CircleShape).background(color.copy(alpha = 0.3f)))
        return
    }
    val transition = rememberInfiniteTransition(label = "mate-btn-spin")
    val rotation by transition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(800, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "mate-btn-spin-r",
    )
    // 简化 spinner：旋转的环（外圈淡色轨道 + 内圈高亮弧，用两段 Box 近似）。
    Box(modifier = Modifier.size(size).rotate(rotation)) {
        Box(Modifier.size(size).clip(CircleShape).background(color.copy(alpha = 0.3f)))
    }
}

/**
 * 角标（badge）。
 *
 * min-w 18，h18，padding 0 5，radius 9，bg brand，#fff，caption semibold（>99 显示 99+）。
 */
@Composable
private fun MateBadge(count: Int) {
    val display = if (count > 99) "99+" else count.toString()
    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(BrandColor)
            .padding(horizontal = 5.dp, vertical = 1.dp)
            .height(16.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(display, color = Color.White, fontSize = 11.sp, fontWeight = FontWeight.SemiBold)
    }
}

File: lib/widgets/mate_button.dart
import 'package:flutter/material.dart';

import '../app/theme/mate_theme.dart';
import 'mate_icon.dart';

// =============================================================================
// MateButton —— 五变体统一按钮（对标 CMP mate/MateButton.kt v2）。
//
// - primary：品牌渐变底（brandHover→brand），白字，radius 8，h36，品牌色柔影；
//   hover 降透明；pressed→brandActive 纯色；danger→error 纯色
// - soft：brandLighter 浅蓝底，brand 字，radius 8，h36；hover→brand100
// - text：透明底，brand 字，radius 5；hover→brandLighter；danger 字色 error
// - icon：透明，32×32 正圆形；hover→bgFill + textPrimary；danger 字色 error
// - iconText：透明，h36，radius 8；hover→bgFill
//
// hover 仅背景色过渡，无 ripple；loading 时 primary 显 spinner，其余图标 spin + 禁用点击。
// =============================================================================

/// 按钮形态（v2：主按钮 / 软色 / 文字 / 图标 / 图标文字）。
enum MateButtonVariant {
  /// 品牌渐变主按钮。
  primary,

  /// 软色按钮（浅蓝底）。
  soft,

  /// 文字按钮。
  text,

  /// 纯图标圆形按钮。
  icon,

  /// 图标 + 文字按钮。
  iconText,
}

/// 统一按钮（macOS 风格，无水波纹）。
///
/// 示例：
/// ```dart
/// MateButton(
///   label: '确定',
///   variant: MateButtonVariant.primary,
///   onClick: () => print('clicked'),
/// )
/// ```
class MateButton extends StatefulWidget {
  /// 按钮文字（icon 变体可空）。
  final String? label;

  /// 按钮形态。
  final MateButtonVariant variant;

  /// 点击回调。
  final VoidCallback? onClick;

  /// 图标 name（可选；primary/text 用 14px，iconText/soft 16px，icon 18px）。
  final String? icon;

  /// 危险态（红）。
  final bool danger;

  /// 禁用。
  final bool disabled;

  /// 加载中（primary 显 spinner，其余图标 spin + 禁用点击）。
  final bool loading;

  /// 100% 宽（仅 primary/text/soft/iconText 生效）。
  final bool fullWidth;

  /// 角标数字（>0 显示，仅 icon/iconText）。
  final int badge;

  /// 自定义高度（null 用变体默认）。
  final double? height;

  const MateButton({
    super.key,
    this.label,
    this.variant = MateButtonVariant.primary,
    this.onClick,
    this.icon,
    this.danger = false,
    this.disabled = false,
    this.loading = false,
    this.fullWidth = false,
    this.badge = 0,
    this.height,
  });

  @override
  State<MateButton> createState() => _MateButtonState();
}

class _MateButtonState extends State<MateButton> {
  bool _hovered = false;
  bool _pressed = false;

  bool get _effectiveDisabled => widget.disabled || widget.loading;

  @override
  Widget build(BuildContext context) {
    final colors = MateTheme.colorsOf(context);
    final typography = MateTheme.typographyOf(context).button;
    final metrics = MateTheme.metricsOf(context).button;

    // ---- 形状：icon 为正圆，text 为 radius-sm，其余 radius-md ----
    final BoxShape shape;
    final BorderRadius borderRadius;
    switch (widget.variant) {
      case MateButtonVariant.icon:
        shape = BoxShape.circle;
        borderRadius = BorderRadius.zero;
      case MateButtonVariant.text:
        shape = BoxShape.rectangle;
        borderRadius = BorderRadius.circular(metrics.textRadius);
      case MateButtonVariant.soft:
        shape = BoxShape.rectangle;
        borderRadius = BorderRadius.circular(metrics.softRadius);
      case MateButtonVariant.iconText:
        shape = BoxShape.rectangle;
        borderRadius = BorderRadius.circular(metrics.iconTextRadius);
      case MateButtonVariant.primary:
        shape = BoxShape.rectangle;
        borderRadius = BorderRadius.circular(metrics.primaryRadius);
    }

    // ---- 背景与文字色按变体 × 状态组合计算（无 ripple，纯背景过渡） ----
    // primary 返回的 bgColor 仅作 danger/disabled 兜底；正常态用渐变刷。
    final Color bgColor;
    final Color contentColor;
    switch (widget.variant) {
      case MateButtonVariant.primary:
        if (_effectiveDisabled) {
          bgColor = colors.border;
          contentColor = colors.buttonDisabledPrimaryText;
        } else if (widget.danger) {
          bgColor = _pressed
              ? colors.error.withAlpha((metrics.dangerPressedAlpha * 255).round())
              : colors.error;
          contentColor = colors.buttonDangerText;
        } else {
          bgColor = colors.brand;
          contentColor = colors.buttonPrimaryText;
        }
      case MateButtonVariant.soft:
        if (_effectiveDisabled) {
          bgColor = colors.bgFill;
          contentColor = colors.textPlaceholder;
        } else {
          bgColor = _hovered ? colors.brand100 : colors.brandLighter;
          contentColor = colors.brand;
        }
      case MateButtonVariant.text:
        if (_effectiveDisabled) {
          bgColor = Colors.transparent;
          contentColor = colors.textPlaceholder;
        } else if (widget.danger) {
          bgColor = _hovered ? colors.brandLighter : Colors.transparent;
          contentColor = colors.error;
        } else {
          bgColor = _hovered ? colors.brandLighter : Colors.transparent;
          contentColor = colors.brand;
        }
      case MateButtonVariant.icon:
        if (_effectiveDisabled) {
          bgColor = Colors.transparent;
          contentColor = colors.textPlaceholder;
        } else if (_hovered) {
          bgColor = colors.bgFill;
          contentColor = colors.textPrimary;
        } else {
          bgColor = Colors.transparent;
          contentColor = widget.danger ? colors.error : colors.textSecondary;
        }
      case MateButtonVariant.iconText:
        if (_effectiveDisabled) {
          bgColor = Colors.transparent;
          contentColor = colors.textPlaceholder;
        } else if (_hovered) {
          bgColor = colors.bgFill;
          contentColor = colors.textPrimary;
        } else {
          bgColor = Colors.transparent;
          contentColor = colors.textPrimary;
        }
    }

    // ---- 高度 / 图标尺寸 / 水平内边距 ----
    final resolvedHeight = widget.height ??
        switch (widget.variant) {
          MateButtonVariant.icon => metrics.iconButtonSize,
          MateButtonVariant.text => metrics.textHeight,
          MateButtonVariant.soft => metrics.softHeight,
          MateButtonVariant.iconText => metrics.iconTextHeight,
          MateButtonVariant.primary => metrics.primaryHeight,
        };
    final iconSize = switch (widget.variant) {
      MateButtonVariant.icon => metrics.iconVariantIconSize,
      MateButtonVariant.iconText => metrics.iconTextVariantIconSize,
      MateButtonVariant.soft => metrics.softVariantIconSize,
      MateButtonVariant.primary => widget.icon != null
          ? metrics.primaryVariantIconSize
          : metrics.primaryWithoutIconSize,
      MateButtonVariant.text => widget.icon != null
          ? metrics.textVariantIconSize
          : metrics.textWithoutIconSize,
    };
    final horizontalPadding = switch (widget.variant) {
      MateButtonVariant.icon => metrics.iconHorizontalPadding,
      MateButtonVariant.iconText => metrics.iconTextHorizontalPadding,
      MateButtonVariant.text => metrics.textHorizontalPadding,
      MateButtonVariant.soft => metrics.softHorizontalPadding,
      MateButtonVariant.primary => metrics.primaryHorizontalPadding,
    };
    final labelStyle = switch (widget.variant) {
      MateButtonVariant.primary => typography.primaryLabel,
      MateButtonVariant.soft => typography.softLabel,
      MateButtonVariant.text => typography.textLabel,
      MateButtonVariant.iconText => typography.iconTextLabel,
      MateButtonVariant.icon => typography.iconTextLabel,
    };

    // primary 正常态（非 danger 非禁用）用品牌渐变 + 品牌色柔影，pressed 转 brandActive 纯色。
    final useGradient = widget.variant == MateButtonVariant.primary &&
        !widget.danger &&
        !_effectiveDisabled;
    final pressedSolid = useGradient && _pressed;

    final decoration = BoxDecoration(
      shape: shape,
      borderRadius: shape == BoxShape.circle ? null : borderRadius,
      color: useGradient && !pressedSolid ? null : bgColor,
      gradient: useGradient && !pressedSolid
          ? LinearGradient(
              colors: colors.brandGradient,
              begin: Alignment.topLeft,
              end: Alignment.bottomRight,
            )
          : null,
      boxShadow: useGradient
          ? [
              BoxShadow(
                color: colors.brand
                    .withAlpha((metrics.primaryShadowAlpha * 255).round()),
                blurRadius: metrics.primaryShadowElevation,
                offset: Offset(0, metrics.primaryShadowElevation / 2),
              ),
            ]
          : null,
    );

    // ---- 内容行 ----
    final children = <Widget>[];
    if (widget.loading && widget.variant == MateButtonVariant.primary) {
      children.add(_MateButtonSpinner(
        size: metrics.loadingSpinnerSize,
        color: contentColor,
        trackAlpha: metrics.spinnerTrackAlpha,
        duration: Duration(milliseconds: metrics.spinnerRotationDurationMillis),
        reducedMotion: MateTheme.reducedMotionOf(context),
      ));
      if (widget.label != null) {
        children.add(SizedBox(width: metrics.loadingLabelSpacing));
      }
    } else if (widget.icon != null) {
      final spin = widget.loading && widget.variant != MateButtonVariant.primary;
      children.add(MateIcon(
        name: widget.icon!,
        size: iconSize,
        tint: contentColor,
        spin: spin,
      ));
      if (widget.label != null) {
        children.add(SizedBox(width: metrics.iconLabelSpacing));
      }
    }
    if (widget.label != null) {
      children.add(Text(
        widget.label!,
        style: labelStyle.copyWith(color: contentColor),
      ));
    }
    if (widget.badge > 0 && widget.variant == MateButtonVariant.iconText) {
      children.add(Padding(
        padding: EdgeInsets.only(left: metrics.badgeStartPadding),
        child: _MateBadge(count: widget.badge),
      ));
    }

    Widget button = AnimatedContainer(
      duration: const Duration(milliseconds: 150),
      curve: Curves.easeOut,
      height: resolvedHeight,
      width: widget.variant == MateButtonVariant.icon
          ? metrics.iconButtonSize
          : (widget.fullWidth ? double.infinity : null),
      padding: EdgeInsets.symmetric(horizontal: horizontalPadding),
      decoration: decoration,
      child: Opacity(
        opacity: _effectiveDisabled ? metrics.disabledAlpha : 1,
        child: Row(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: children,
        ),
      ),
    );

    // icon 变体的 badge 置于 32×32 命中区外侧（避免圆形容器内溢出）
    if (widget.badge > 0 && widget.variant == MateButtonVariant.icon) {
      button = Row(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          button,
          Padding(
            padding: EdgeInsets.only(left: metrics.badgeStartPadding),
            child: _MateBadge(count: widget.badge),
          ),
        ],
      );
    }

    return MouseRegion(
      cursor: _effectiveDisabled
          ? SystemMouseCursors.basic
          : SystemMouseCursors.click,
      onEnter: (_) => _setHover(true),
      onExit: (_) => _setHover(false),
      child: GestureDetector(
        onTapDown: _effectiveDisabled ? null : (_) => _setPressed(true),
        onTapUp: _effectiveDisabled ? null : (_) => _setPressed(false),
        onTapCancel: _effectiveDisabled ? null : () => _setPressed(false),
        onTap: _effectiveDisabled ? null : widget.onClick,
        behavior: HitTestBehavior.opaque,
        child: button,
      ),
    );
  }

  void _setHover(bool v) {
    if (_hovered != v) setState(() => _hovered = v);
  }

  void _setPressed(bool v) {
    if (_pressed != v) setState(() => _pressed = v);
  }
}

/// primary 加载态的小 spinner（16×16，0.8s 线性旋转；减少动态时降级为静态半透明圆）。
class _MateButtonSpinner extends StatefulWidget {
  final double size;
  final Color color;
  final double trackAlpha;
  final Duration duration;
  final bool reducedMotion;

  const _MateButtonSpinner({
    required this.size,
    required this.color,
    required this.trackAlpha,
    required this.duration,
    required this.reducedMotion,
  });

  @override
  State<_MateButtonSpinner> createState() => _MateButtonSpinnerState();
}

class _MateButtonSpinnerState extends State<_MateButtonSpinner>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this, duration: widget.duration);
    if (!widget.reducedMotion) _controller.repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final spinner = CustomPaint(
      size: Size.square(widget.size),
      painter: _ArcSpinnerPainter(
        color: widget.color,
        trackAlpha: widget.trackAlpha,
      ),
    );
    if (widget.reducedMotion) return spinner;
    return AnimatedBuilder(
      animation: _controller,
      builder: (_, child) => Transform.rotate(
        angle: _controller.value * 2 * 3.141592653589793,
        child: child,
      ),
      child: spinner,
    );
  }
}

/// 圆弧 spinner 绘制（淡色轨道整圈 + 高亮 90° 弧）。
class _ArcSpinnerPainter extends CustomPainter {
  final Color color;
  final double trackAlpha;

  _ArcSpinnerPainter({required this.color, required this.trackAlpha});

  @override
  void paint(Canvas canvas, Size size) {
    const strokeWidth = 2.0;
    final rect = Offset.zero & size;
    final trackPaint = Paint()
      ..color = color.withAlpha((trackAlpha * 255).round())
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;
    final arcPaint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round;
    canvas.drawCircle(size.center(Offset.zero), size.width / 2 - strokeWidth / 2, trackPaint);
    canvas.drawArc(
      rect.deflate(strokeWidth / 2),
      -3.141592653589793 / 2,
      3.141592653589793 / 2,
      false,
      arcPaint,
    );
  }

  @override
  bool shouldRepaint(covariant _ArcSpinnerPainter oldDelegate) =>
      oldDelegate.color != color || oldDelegate.trackAlpha != trackAlpha;
}

/// 角标（badge）。
///
/// h16，padding 0/5，pill 圆角，bg brand，白字 10 semibold（>99 显示 99+）。
class _MateBadge extends StatelessWidget {
  final int count;

  const _MateBadge({required this.count});

  @override
  Widget build(BuildContext context) {
    final colors = MateTheme.colorsOf(context);
    final metrics = MateTheme.metricsOf(context).button;
    final display = count > 99 ? '99+' : count.toString();
    return Container(
      height: metrics.badgeHeight,
      padding: EdgeInsets.symmetric(horizontal: metrics.badgeHorizontalPadding),
      decoration: BoxDecoration(
        color: colors.brand,
        borderRadius: BorderRadius.circular(metrics.badgeHeight / 2),
      ),
      alignment: Alignment.center,
      child: Text(
        display,
        style: MateTheme.typographyOf(context)
            .button
            .badgeLabel
            .copyWith(color: colors.buttonBadgeText),
      ),
    );
  }
}
